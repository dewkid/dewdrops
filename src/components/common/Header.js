import React, { useState, useEffect } from 'react';
import SelectInput from './SelectInput';
import { connect } from 'react-redux';
import * as userActions from '../../redux/actions/userActions';
import * as loginActions from '../../redux/actions/currentUserActions';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';

export function Header({ currentUser, users, actions }) {
  const navbarStyle = { background: '#bdffa9' };
  const [login, setLogin] = useState(currentUser);

  useEffect(() => {
    if (users.length === 0) {
      actions.loadUsers().catch(error => {
        alert('Loading users failed' + error);
      });
    }
  }, []);

  function handleChange(event) {
    var user = users.find(user => user.id == event.target.value);
    actions.loginUser(user);
    setLogin({ ...user, loggedIn: true });
  }

  function doLogout() {
    actions.logoutUser();
    setLogin({ loggedIn: false });
  }

  return (
    <nav className='navbar navbar-light' style={navbarStyle}>
      <a className='navbar-brand justify-content-start' href='/'>
        <img src='../../../assets/logo.png' height='50' alt='' />
      </a>

      <a className='nav-link active' href='/'>
        Home
      </a>
      <a className='nav-link' href='/about'>
        About
      </a>
      {login && login.loggedIn ? (
        <div className='row'>
          <div className='col'>
            <h5>{login.name}</h5>
          </div>
          <div className='col'>
            <button onClick={() => doLogout()}>Logout</button>
          </div>
        </div>
      ) : (
        <SelectInput
          name='userid'
          defaultOption={login.name || 'Select User'}
          options={users
            .sort((a, b) => a.name.localeCompare(b.name))
            .map(user => ({
              value: user.id,
              text: user.name
            }))}
          onChange={handleChange}
        />
      )}
    </nav>
  );
}

Header.propTypes = {
  users: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired,
  currentUser: PropTypes.object.isRequired,
  loginid: PropTypes.number
};

function mapStateToProps(state) {
  // be specific which data is exposed, because react will re-render all data it sees here
  return {
    currentUser: state.currentUser,
    users: state.users,
    loading: state.apiCallsInProgress > 0
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: {
      loadUsers: bindActionCreators(userActions.loadUsers, dispatch),
      loginUser: bindActionCreators(loginActions.loginUser, dispatch),
      logoutUser: bindActionCreators(loginActions.logoutUser, dispatch)
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Header);
