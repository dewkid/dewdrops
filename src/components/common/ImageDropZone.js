import React, { useState, useCallback, useMemo } from 'react';
import { useDropzone } from 'react-dropzone';
import PropTypes from 'prop-types';

const baseStyle = {
  flex: 1,
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
  padding: '20px',
  height: '200px',
  borderWidth: 2,
  borderRadius: 2,
  borderColor: '#eeeeee',
  borderStyle: 'dashed',
  backgroundColor: '#fafafa',
  color: '#bdbdbd',
  outline: 'none',
  transition: 'border .24s ease-in-out'
};

const activeStyle = {
  borderColor: '#2196f3'
};

const acceptStyle = {
  borderColor: '#00e676'
};

const rejectStyle = {
  borderColor: '#ff1744'
};

function ImageDropzone({ onFilesDrop, name, label, error }) {
  const [file, setFile] = useState({});
  const onDrop = useCallback(acceptedFiles => {
    previewDrop(acceptedFiles);
    onFilesDrop(acceptedFiles);
  }, []);
  const {
    getRootProps,
    getInputProps,
    isDragActive,
    isDragAccept,
    isDragReject
  } = useDropzone({
    onDrop,
    multiple: false,
    accept: 'image/*',
    autoProcessQueue: false
  });

  function previewDrop(acceptedFiles) {
    if (acceptedFiles.length > 0) {
      let firstFile = acceptedFiles[0];
      firstFile.preview = URL.createObjectURL(firstFile);
      setFile(firstFile);
      console.log('name:' + firstFile.name + ' preview:' + firstFile.preview);
    }
  }

  const style = useMemo(
    () => ({
      ...baseStyle,
      ...(isDragActive ? activeStyle : {}),
      ...(isDragAccept ? acceptStyle : {}),
      ...(isDragReject ? rejectStyle : {})
    }),
    [isDragActive, isDragReject]
  );

  return (
    <div>
      <label
        htmlFor={name}
        style={{ marginTop: 20, fontSize: 18, fontWeight: 'bold' }}
      >
        {label}
      </label>
      <div {...getRootProps({ style })}>
        <input {...getInputProps()} />
        {isDragActive ? (
          <p>Drop the photo here ...</p>
        ) : (
          <p>Drag-n-drop photo here, or click to select photo</p>
        )}
        {file.preview ? (
          <img
            alt='Preview'
            key={file.preview}
            src={file.preview}
            style={{ display: 'inline', width: 'auto', height: '100px' }}
          />
        ) : (
          <div />
        )}
      </div>
      {error && <div className='alert alert-danger'>{error}</div>}
    </div>
  );
}
ImageDropzone.propTypes = {
  onFilesDrop: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  error: PropTypes.string
};
export default ImageDropzone;
