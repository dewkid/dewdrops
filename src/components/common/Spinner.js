import React from 'react';
import './Spinner.css';

const Spinner = () => {
  return (
    <>
      <div className='kart-loader'>
        <div className='sheath'>
          <div className='segment'></div>
        </div>
        <div className='sheath'>
          <div className='segment'></div>
        </div>
        <div className='sheath'>
          <div className='segment'></div>
        </div>
        <div className='sheath'>
          <div className='segment'></div>
        </div>
        <div className='sheath'>
          <div className='segment'></div>
        </div>
        <div className='sheath'>
          <div className='segment'></div>
        </div>
        <div className='sheath'>
          <div className='segment'></div>
        </div>
        <div className='sheath'>
          <div className='segment'></div>
        </div>
        <div className='sheath'>
          <div className='segment'></div>
        </div>
        <div className='sheath'>
          <div className='segment'></div>
        </div>
        <div className='sheath'>
          <div className='segment'></div>
        </div>
        <div className='sheath'>
          <div className='segment'></div>
        </div>
      </div>
    </>
  );
};

export default Spinner;
