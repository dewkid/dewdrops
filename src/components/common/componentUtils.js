const prodImagesPath = ''; // Will point to AWS
const devImagesPath = '../../../tools/imageData';

function getImagePath(dewdrop) {
  const storedImagesPath =
    process.env.NODE_ENV === 'production' ? prodImagesPath : devImagesPath;

  return storedImagesPath + '/' + dewdrop.imageName;
}

export default getImagePath;
