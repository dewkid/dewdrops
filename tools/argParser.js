function setPassword(argv) {
  const args = argv.filter(value => value.startsWith('--password='));
  if (args.length == 1) {
    let passwordArg = args[0];
    process.env.DB_PASSWORD = passwordArg.substring(11);
  }
  if (!process.env.DB_PASSWORD || process.env.DB_PASSWORD.length == 0) {
    console.log('--password=xxxx argument is required');
    process.exit(1);
  }
}

module.exports = {
  setPassword
};
