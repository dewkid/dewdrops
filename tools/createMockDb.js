/* eslint-disable no-console */
const util = require('util');
const fs = require('fs');
const rimraf = require('rimraf');
const path = require('path');
const { dewdrops, comments, users } = require('./mockData');
const copyFilePromise = util.promisify(fs.copyFile);
const { setPassword } = require('./argParser');

setPassword(process.argv);

const { User, Dewdrop, Comment, connection } = require('./dbModels');

function copyFiles(srcDir, destDir, files) {
  return Promise.all(
    files.map(f => {
      return copyFilePromise(path.join(srcDir, f), path.join(destDir, f));
    })
  );
}

const imageDataPath = path.join(__dirname, 'imageData');
const mockImageDataPath = path.join(__dirname, 'mockImages');

// remove the image directory
rimraf.sync(imageDataPath);

// recreate image directory and copy mock images
if (!fs.existsSync(imageDataPath)) fs.mkdirSync(imageDataPath);
copyFiles(mockImageDataPath, imageDataPath, [
  'bob.jpg',
  'character.jpg',
  'ski.jpg',
  'sunset.jpg',
  'void.jpg'
])
  .then(() => {
    console.log('done');
    setupPostgresTables();
  })
  .catch(err => {
    console.log(err);
  });

// Testing

function testConnection() {
  connection
    .authenticate()
    .then(() => {
      console.log('Connection has been established successfully.');
    })
    .catch(err => {
      console.error('Unable to connect to the database:', err);
    });
}

async function createTableData() {
  await Promise.all(
    users.map(user => {
      delete user.id;
      User.create(user);
    })
  );
  await Promise.all(
    dewdrops.map(dewdrop => {
      delete dewdrop.id;
      Dewdrop.create(dewdrop);
    })
  );
  await Promise.all(
    comments.map(comment => {
      delete comment.id;
      Comment.create(comment);
    })
  );
}

function setupPostgresTables() {
  testConnection();
  connection.sync({ force: true }).then(() => {
    console.log('db tables created');
    createTableData().then(() => {
      console.log('Mock records created');

      Dewdrop.findAll().then(dewdrops => {
        dewdrops.map(dewdrop => console.log(dewdrop.imageName));
      });
    });
  });
}
