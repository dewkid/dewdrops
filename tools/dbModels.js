const Sequelize = require('sequelize');

const connection = new Sequelize(
  'dewdropsweb',
  'postgres',
  process.env.DB_PASSWORD,
  {
    host: 'localhost',
    dialect: 'postgres' /* one of 'mysql' | 'mariadb' | 'postgres' | 'mssql' */,
    logging: false
  }
);

const User = connection.define(
  'user',
  {
    // attributes
    name: {
      type: Sequelize.STRING,
      allowNull: false
    }
  },
  {
    // options
  }
);

const Dewdrop = connection.define(
  'dewdrop',
  {
    // attributes
    userid: {
      type: Sequelize.INTEGER,
      allowNull: false
    },
    imageName: {
      type: Sequelize.STRING,
      allowNull: false
    },
    caption: {
      type: Sequelize.STRING,
      allowNull: false
    }
  },
  {
    // options
  }
);

const Comment = connection.define(
  'comment',
  {
    // attributes
    dewdropid: {
      type: Sequelize.INTEGER,
      allowNull: false
    },
    userid: {
      type: Sequelize.INTEGER,
      allowNull: false
    },
    commentText: {
      type: Sequelize.STRING,
      allowNull: false
    }
  },
  {
    // options
  }
);

// Using CommonJS style export so we can consume via Node (without using Babel-node)
module.exports = {
  connection,
  User,
  Dewdrop,
  Comment
};
