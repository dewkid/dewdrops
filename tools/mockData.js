const dewdrops = [
  {
    id: 1,
    userid: 1,
    imageName: 'character.jpg',
    caption: 'My next D&D character!'
  },
  {
    id: 2,
    userid: 2,
    imageName: 'sunset.jpg',
    caption: 'Check out this sunset'
  },
  {
    id: 3,
    userid: 1,
    imageName: 'ski.jpg',
    caption: 'Killer shot of me hitting the slopes.'
  },
  {
    id: 4,
    userid: 3,
    imageName: 'bob.jpg',
    caption: 'THIS guy!'
  }
];

const comments = [
  { id: 1, dewdropid: 1, userid: 2, commentText: 'Sweet!' },
  { id: 2, dewdropid: 1, userid: 3, commentText: "What's his name?" },
  {
    id: 3,
    dewdropid: 2,
    userid: 4,
    commentText: 'Where did you take that?  Is that near your place?'
  },
  { id: 4, dewdropid: 1, userid: 1, commentText: "I'm thinking Garr or Frank" },
  { id: 5, dewdropid: 1, userid: 3, commentText: 'Frank?' },
  { id: 6, dewdropid: 2, userid: 1, commentText: 'Yeah, where is that?' },
  { id: 7, dewdropid: 1, userid: 4, commentText: 'LOL - how about Jeeves' },
  { id: 8, dewdropid: 1, userid: 1, commentText: 'LOL' },
  { id: 9, dewdropid: 2, userid: 2, commentText: 'Not my pic, just liked it.' },
  {
    id: 10,
    dewdropid: 3,
    userid: 2,
    commentText: 'Right, like you could do that.'
  },
  { id: 11, dewdropid: 2, userid: 3, commentText: 'Lame!' },
  { id: 12, dewdropid: 3, userid: 1, commentText: "That's me!" },
  { id: 13, dewdropid: 2, userid: 1, commentText: 'I knew it!' },
  {
    id: 14,
    dewdropid: 4,
    userid: 4,
    commentText: 'Is that what you are wearing for the party?'
  },
  { id: 15, dewdropid: 4, userid: 3, commentText: 'Thinking about it...' },
  {
    id: 16,
    dewdropid: 4,
    userid: 3,
    commentText: 'You totally should, that would be hilarious.'
  }
];

const users = [
  { id: 1, name: 'Robin', username: 'robin', password: 'robin' },
  { id: 2, name: 'Dave', username: 'dave', password: 'dave' },
  { id: 3, name: 'Ross', username: 'ross', password: 'ross' },
  { id: 4, name: 'Ryan', username: 'ryan', password: 'ryan' },
  { id: 5, name: 'Lance', username: 'lance', password: 'lance' },
  { id: 6, name: 'Paul', username: 'paul', password: 'lance' },
  { id: 7, name: 'Nathaniel', username: 'nathaniel', password: 'nathaniel' },
  { id: 8, name: 'Chris', username: 'chris', password: 'chris' },
  { id: 9, name: 'Sam', username: 'sam', password: 'sam' }
];

const newDewdrop = {
  id: null,
  userid: null,
  imageName: '',
  caption: ''
};

const newComment = {
  id: null,
  dewdropid: null,
  userid: null,
  commentText: ''
};

// Using CommonJS style export so we can consume via Node (without using Babel-node)
module.exports = {
  newDewdrop,
  newComment,
  dewdrops,
  comments,
  users
};
