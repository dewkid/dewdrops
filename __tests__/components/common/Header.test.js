import React from 'react';
import renderer from 'react-test-renderer';
// Component to be tested
import { Header } from '../../../src/components/common/Header';
import { users } from '../../../tools/mockData';

it('Header snapshot', () => {
  const tree = renderer.create(
    <Header currentUser={{ id: 1, name: 'test' }} users={users} actions={{}} />
  );

  expect(tree).toMatchSnapshot();
});
