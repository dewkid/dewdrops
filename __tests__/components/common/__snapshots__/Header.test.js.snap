// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header snapshot 1`] = `
<nav
  className="navbar navbar-light"
  style={
    Object {
      "background": "#bdffa9",
    }
  }
>
  <a
    className="navbar-brand justify-content-start"
    href="/"
  >
    <img
      alt=""
      height="50"
      src="../../../assets/logo.png"
    />
  </a>
  <a
    className="nav-link active"
    href="/"
  >
    Home
  </a>
  <a
    className="nav-link"
    href="/about"
  >
    About
  </a>
  <div
    className="form-group"
  >
    
    <div
      className="field"
    >
      <select
        className="form-control"
        name="userid"
        onChange={[Function]}
      >
        <option
          value=""
        >
          test
        </option>
        <option
          value={8}
        >
          Chris
        </option>
        <option
          value={2}
        >
          Dave
        </option>
        <option
          value={5}
        >
          Lance
        </option>
        <option
          value={7}
        >
          Nathaniel
        </option>
        <option
          value={6}
        >
          Paul
        </option>
        <option
          value={1}
        >
          Robin
        </option>
        <option
          value={3}
        >
          Ross
        </option>
        <option
          value={4}
        >
          Ryan
        </option>
        <option
          value={9}
        >
          Sam
        </option>
      </select>
    </div>
  </div>
</nav>
`;
